Spi flash_map for intel_s1000

rgundi opened the pull request:
Enable flash_map support of intel_s1000. It makes use of the generic spi nor flash driver.

rgundi committed:
drivers: flash: spi_nor: Enable Flash Page Layout and Map

This patch enables the support for FLASH_PAGE_LAYOUT and
FLASH_MAP for the generic spin nor flash driver.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests: intel_s1000: flash_map: Enable tests for flash_map

This patchset introduces some tests for the flash_map and
flash_page_layout features.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
dts: intel_s1000_crb: Enable flash page layout support

This patchset introduces partitioning of flash in terms of blocks
(flash page layout) for s1000. This partitioning is currently to
be used for mcubootloader.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
@galak: Help needed. I see the below warnings with this patchset. Not sure how to fix it. It doesn't seem to have any effect though and things work fine.
intel_s1000_crb.dts_compiled: Warning (spi_bus_bridge): /soc/spi@e000: incorrect #size-cells for SPI bus
intel_s1000_crb.dts_compiled: Warning (spi_bus_reg): Failed prerequisite 'spi_bus_bridge'

nashif commented on drivers/flash/Kconfig.nor:
```diff
@@ -9,6 +9,8 @@ menuconfig SPI_NOR
 	prompt "SPI NOR Flash"
 	select FLASH_HAS_DRIVER_ENABLED
+	select FLASH_HAS_PAGE_LAYOUT
+	select FLASH_PAGE_LAYOUT
```
these selects here are wrong and causing kconfig to go into a loop and breaks documentation.

rgundi replied:
Have addressed this and shippable is passing now.

galak commented:
> @galak: Help needed. I see the below warnings with this patchset. Not sure how to fix it. It doesn't seem to have any effect though and things work fine.
> intel_s1000_crb.dts_compiled: Warning (spi_bus_bridge): /soc/spi@e000: incorrect #size-cells for SPI bus
> intel_s1000_crb.dts_compiled: Warning (spi_bus_reg): Failed prerequisite 'spi_bus_bridge'

Can you describe the actual flash on the S1000.

rgundi commented:
> > @galak: Help needed. I see the below warnings with this patchset. Not sure how to fix it. It doesn't seem to have any effect though and things work fine.
> > intel_s1000_crb.dts_compiled: Warning (spi_bus_bridge): /soc/spi@e000: incorrect #size-cells for SPI bus
> > intel_s1000_crb.dts_compiled: Warning (spi_bus_reg): Failed prerequisite 'spi_bus_bridge'
> 
> Can you describe the actual flash on the S1000.

It is a Macronix flash of 512MB sitting on the SPI bus. It has a page of 256 bytes and sector of 4kB. Erase is possible 32kB or 64kB at a time. Anything else you are expecting?

sathishkuttan commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts:
```diff
@@ -34,11 +35,52 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		/* The flash is 512MB in size. Each sector is 4kB and
```
The flash size is 512Mb (as it Mega *bits*). In MB units, it is 64MB.

rgundi replied:
Oops. Will correct. Thanks for pointing out.

galak commented:
> It is a Macronix flash of 512MB sitting on the SPI bus. It has a page of 256 bytes and sector of 4kB. Erase is possible 32kB or 64kB at a time. Anything else you are expecting?

do you need to know the flash size, etc at build time?

rgundi commented:
> > It is a Macronix flash of 512MB sitting on the SPI bus. It has a page of 256 bytes and sector of 4kB. Erase is possible 32kB or 64kB at a time. Anything else you are expecting?
> 
> do you need to know the flash size, etc at build time?

Yes. It is required for CONFIG_FLASH_PAGE_LAYOUT.

nashif commented:
You still have Kconfig issues 

```
Traceback (most recent call last):
  File "../ci-tools/scripts/check_compliance.py", line 593, in <module>
    main()
  File "../ci-tools/scripts/check_compliance.py", line 570, in main
    test.run()
  File "../ci-tools/scripts/check_compliance.py", line 154, in run
    for warning in kconfiglib.Kconfig().warnings:
  File "/home/nashif/Work/github/zephyr/scripts/kconfig/kconfiglib.py", line 997, in __init__
    check_dep_loop_sym(sym, False)
  File "/home/nashif/Work/github/zephyr/scripts/kconfig/kconfiglib.py", line 6010, in _check_dep_loop_sym
    return _found_dep_loop(loop, sym)
  File "/home/nashif/Work/github/zephyr/scripts/kconfig/kconfiglib.py", line 6112, in _found_dep_loop
    raise KconfigError(msg)
kconfiglib.KconfigError:
Dependency loop
===============

SPI (defined at boards/arm/96b_carbon/Kconfig.defconfig:56, boards/arm/disco_l475_iot1/Kconfig.defconfig:128, boards/arm/disco_l475_iot1/Kconfig.defconfig:150, boards/x86/galileo/Kconfig.defconfig:19, boards/x86/quark_se_c1000_devboard/Kconfig.defconfig:13, soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk64f12:60, soc/arm/nxp_kinetis/kwx/Kconfig.defconfig.mkw2xd512:65, drivers/spi/Kconfig:12), with definition...

config SPI
	bool
	default "y" if BT && BOARD_96B_CARBON
	depends on BT && BOARD_96B_CARBON

config SPI
	bool
	default "y" if BT && BOARD_DISCO_L475_IOT1
	depends on BT && BOARD_DISCO_L475_IOT1

config SPI
	bool
	default "y" if WIFI && BOARD_DISCO_L475_IOT1
	depends on WIFI && BOARD_DISCO_L475_IOT1

config SPI
	bool
	default "y" if BOARD_GALILEO
	depends on BOARD_GALILEO

config SPI
	bool
	default "y" if IEEE802154_CC2520 && BOARD_QUARK_SE_C1000_DEVBOARD
	depends on IEEE802154_CC2520 && BOARD_QUARK_SE_C1000_DEVBOARD

config SPI
	bool
	default "n" if SOC_MK64F12 && SOC_SERIES_KINETIS_K6X
	depends on SOC_MK64F12 && SOC_SERIES_KINETIS_K6X

config SPI
	bool
	default "y" if (SOC_MKW22D5 || SOC_MKW24D5) && SOC_SERIES_KINETIS_KWX
	depends on (SOC_MKW22D5 || SOC_MKW24D5) && SOC_SERIES_KINETIS_KWX

menuconfig SPI
	bool
	prompt "SPI hardware bus support"
	select HAS_DTS_SPI
	help
	  Enable support for the SPI hardware bus.

...depends on IEEE802154_CC2520 (defined at soc/x86/intel_quark/quark_se/Kconfig.defconfig.quark_se_c1000:16, drivers/ieee802154/Kconfig.cc2520:9), with definition...

config IEEE802154_CC2520
	bool
	default "y" if IEEE802154 && NET_CONFIG_SETTINGS && SOC_QUARK_SE_C1000 && SOC_SERIES_QUARK_SE
	depends on IEEE802154 && NET_CONFIG_SETTINGS && SOC_QUARK_SE_C1000 && SOC_SERIES_QUARK_SE

menuconfig IEEE802154_CC2520
	bool
	prompt "TI CC2520 Driver support" if NETWORKING && IEEE802154
	depends on NETWORKING && IEEE802154

...depends on IEEE802154 (defined at drivers/ieee802154/Kconfig:12), with definition...

menuconfig IEEE802154
	bool
	prompt "IEEE 802.15.4 drivers options"
	default "y" if NET_L2_IEEE802154 || NET_L2_OPENTHREAD

...depends on NET_L2_OPENTHREAD (defined at subsys/net/l2/openthread/Kconfig:12), with definition...

menuconfig NET_L2_OPENTHREAD
	bool
	prompt "OpenThread L2" if FLASH && FLASH_PAGE_LAYOUT && CPLUSPLUS && REBOOT && NETWORKING
	select OPENTHREAD_PLAT if FLASH && FLASH_PAGE_LAYOUT && CPLUSPLUS && REBOOT && NETWORKING
	depends on FLASH && FLASH_PAGE_LAYOUT && CPLUSPLUS && REBOOT && NETWORKING

...depends on FLASH_PAGE_LAYOUT (defined at boards/xtensa/intel_s1000_crb/Kconfig.defconfig:175, drivers/flash/Kconfig:41), with definition...

config FLASH_PAGE_LAYOUT
	bool
	default "y" if SPI && BOARD_INTEL_S1000_CRB
	depends on SPI && BOARD_INTEL_S1000_CRB

config FLASH_PAGE_LAYOUT
	bool
	prompt "API for retrieving the layout of pages" if FLASH_HAS_PAGE_LAYOUT && FLASH
	depends on FLASH_HAS_PAGE_LAYOUT && FLASH
	help
	  Enables API for retrieving the layout of flash memory pages.

(select-related dependencies: (SOC_FLASH_STM32 && SOC_SERIES_STM32F0X && (SOC_SERIES_STM32F0X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32L4X) && SOC_FAMILY_STM32 && FLASH) || (SOC_FLASH_STM32 && SOC_SERIES_STM32F4X && (SOC_SERIES_STM32F0X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32L4X) && SOC_FAMILY_STM32 && FLASH) || (SOC_FLASH_STM32 && SOC_SERIES_STM32L4X && (SOC_SERIES_STM32F0X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32L4X) && SOC_FAMILY_STM32 && FLASH) || (FILE_SYSTEM_NFFS && FILE_SYSTEM))

...depends on FLASH_HAS_PAGE_LAYOUT (defined at boards/xtensa/intel_s1000_crb/Kconfig.defconfig:172, drivers/flash/Kconfig:17), with definition...

config FLASH_HAS_PAGE_LAYOUT
	bool
	default "y" if SPI && BOARD_INTEL_S1000_CRB
	depends on SPI && BOARD_INTEL_S1000_CRB

config FLASH_HAS_PAGE_LAYOUT
	bool
	help
	  This option is enabled when the SoC flash driver supports
	  retrieving the layout of flash memory pages.

(select-related dependencies: (SOC_FLASH_NRF && SOC_FAMILY_NRF && FLASH) || (SOC_FLASH_MCUX && HAS_MCUX && FLASH) || (SOC_FLASH_STM32 && SOC_SERIES_STM32F0X && (SOC_SERIES_STM32F0X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32L4X) && SOC_FAMILY_STM32 && FLASH) || (SOC_FLASH_STM32 && SOC_SERIES_STM32F4X && (SOC_SERIES_STM32F0X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32L4X) && SOC_FAMILY_STM32 && FLASH) || (SOC_FLASH_STM32 && SOC_SERIES_STM32L4X && (SOC_SERIES_STM32F0X || SOC_SERIES_STM32F3X || SOC_SERIES_STM32F4X || SOC_SERIES_STM32L4X) && SOC_FAMILY_STM32 && FLASH) || (SOC_FLASH_SAM0 && SOC_FAMILY_SAM0 && FLASH) || (SOC_FLASH_SAM && SOC_FAMILY_SAM && FLASH) || (TEST_FLASH_DRIVERS && BOARD_QEMU_X86))

...depends again on SPI (defined at boards/arm/96b_carbon/Kconfig.defconfig:56, boards/arm/disco_l475_iot1/Kconfig.defconfig:128, boards/arm/disco_l475_iot1/Kconfig.defconfig:150, boards/x86/galileo/Kconfig.defconfig:19, boards/x86/quark_se_c1000_devboard/Kconfig.defconfig:13, soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk64f12:60, soc/arm/nxp_kinetis/kwx/Kconfig.defconfig.mkw2xd512:65, drivers/spi/Kconfig:12)

```

rgundi commented:
@nashif Fixed the Kconfig issues and things seem to be fine now.

nashif commented:
> @nashif Fixed the Kconfig issues and things seem to be fine now.

thanks

nashif commented on subsys/storage/flash_map/flash_map.c:
```diff
@@ -38,6 +38,9 @@ static const struct driver_map_entry  flash_drivers_map[] = {
 #ifdef CONFIG_SPI_FLASH_W25QXXDV
 	{SPI_FLASH_0_ID, CONFIG_SPI_FLASH_W25QXXDV_DRV_NAME},
 #endif
+#ifdef CONFIG_SPI_NOR_DRV_NAME
```
SPI_NOR_SPI_FREQ_0 is not a Kconfig variable, it is defined in the fixup file only, it needs to be changed to DT_ afaik.

rgundi replied:
Addressed.

rgundi commented:
@galak: The commit 2442382facb34e47171aec4319a316dfd9aad429 (scripts: extract_dts_includes: Support same sensor on different bus) is breaking this PR. I am getting the below error upon doing a CMake. Could you please help?

Traceback (most recent call last):
  File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract_dts_includes.py", line 753, in <module>
    main()
  File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract_dts_includes.py", line 744, in main
    defs = generate_node_definitions()
  File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract_dts_includes.py", line 689, in generate_node_definitions
    extract_node_include_info(reduced, k, k, None)
  File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract_dts_includes.py", line 436, in extract_node_include_info
    y_node = get_binding(root_node_address)
  File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract/globals.py", line 300, in get_binding
    parent_binding = bindings[parent_compat]
KeyError: 'spi,flash'
CMake Error at /home/rgundi/RajZephyr/zephyr/cmake/dts.cmake:193 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  /home/rgundi/RajZephyr/zephyr/cmake/app/boilerplate.cmake:303 (include)
  CMakeLists.txt:2 (include)

rgundi committed:
Kconfig: intel_s1000: Move FIFO_DEPTH to soc Kconfig

FIFO_DEPTH should be in the SoC's Kconfig as this is a
configuration of SPI master controller in SoC.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
DTS: flash: spi_nor: use DT_ for freq and other parameters

Use DT_ instead of CONFIG_ for spi freq, spi bus name, flash device
name and flash base address.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

galak commented:
@rgundi so the DTS needs fixing, what I'm trying to understand is how the SPI flash is exposed into the memory map.  I assume there is some logic in the SPI controller that supports this?  Is this something beyond the base designware SPI controller?

Also, do you know the exact SPI flash model on the board?

galak commented:
So got a few details from @nashif.  So I think we really need the SPI NOR flash code to use JEDEC commands to probe the flash for such information.  If we don't plan on doing that, we should define a set of properties based on the JEDEC specs that match what one can probe (JESD216*)

rgundi commented:
> @galak: The commit [2442382](https://github.com/zephyrproject-rtos/zephyr/commit/2442382facb34e47171aec4319a316dfd9aad429) (scripts: extract_dts_includes: Support same sensor on different bus) is breaking this PR. I am getting the below error upon doing a CMake. Could you please help?
> 
> Traceback (most recent call last):
> File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract_dts_includes.py", line 753, in 
> main()
> File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract_dts_includes.py", line 744, in main
> defs = generate_node_definitions()
> File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract_dts_includes.py", line 689, in generate_node_definitions
> extract_node_include_info(reduced, k, k, None)
> File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract_dts_includes.py", line 436, in extract_node_include_info
> y_node = get_binding(root_node_address)
> File "/home/rgundi/RajZephyr/zephyr/scripts/dts/extract/globals.py", line 300, in get_binding
> parent_binding = bindings[parent_compat]
> KeyError: 'spi,flash'
> CMake Error at /home/rgundi/RajZephyr/zephyr/cmake/dts.cmake:193 (message):
> command failed with return code: 1
> Call Stack (most recent call first):
> /home/rgundi/RajZephyr/zephyr/cmake/app/boilerplate.cmake:303 (include)
> CMakeLists.txt:2 (include)

Are you talking about this @galak ? My immediate concern is fixing this as otherwise cmake fails. 

> @rgundi so the DTS needs fixing, what I'm trying to understand is how the SPI flash is exposed into the memory map. I assume there is some logic in the SPI controller that supports this? Is this something beyond the base designware SPI controller?
> 
> Also, do you know the exact SPI flash model on the board?
MX25UM51245G Macronix flash

galak commented:
So you have something like:

```
    ...
	#size-cells = <1>;

	flash0: flash@0 {
		compatible = "spi,flash";
		label = "MX25UM512";
		reg = <0x00 0x400000>;
		spi-max-frequency = <2000000>;
		erase-block-size = <0x10000>;
		write-block-size = <1>;
	};
};

```

1. size-cells for SPI should be 0
2. reg should be just the chip select, so 'reg = <0>', not 'reg = <0x00 '0x400000>'

If you make those 2 changes, does it pass the error you are seeing?

rgundi commented:
> So you have something like:
> 
> ```
>     ...
> 	#size-cells = <1>;
> 
> 	flash0: flash@0 {
> 		compatible = "spi,flash";
> 		label = "MX25UM512";
> 		reg = <0x00 0x400000>;
> 		spi-max-frequency = <2000000>;
> 		erase-block-size = <0x10000>;
> 		write-block-size = <1>;
> 	};
> };
> ```
> 1. size-cells for SPI should be 0
> 2. reg should be just the chip select, so 'reg = <0>', not 'reg = <0x00 '0x400000>'
> 
> If you make those 2 changes, does it pass the error you are seeing?

0x400000 denotes the size of the flash used here (4 MB). Though the total flash is 64 MB, we do not want to expose the whole flash for whatever reasons. So, that's needed.

Anyway, i tried both the suggestions and the cmake still doesn't work.

rgundi commented:
> > So you have something like:
> > ```
> >     ...
> > 	#size-cells = <1>;
> > 
> > 	flash0: flash@0 {
> > 		compatible = "spi,flash";
> > 		label = "MX25UM512";
> > 		reg = <0x00 0x400000>;
> > 		spi-max-frequency = <2000000>;
> > 		erase-block-size = <0x10000>;
> > 		write-block-size = <1>;
> > 	};
> > };
> > ```
> > 
> > 1. size-cells for SPI should be 0
> > 2. reg should be just the chip select, so 'reg = <0>', not 'reg = <0x00 '0x400000>'
> > 
> > If you make those 2 changes, does it pass the error you are seeing?
> 
> 0x400000 denotes the size of the flash used here (4 MB). Though the total flash is 64 MB, we do not want to expose the whole flash for whatever reasons. So, that's needed.
> 
> Anyway, i tried both the suggestions and the cmake still doesn't work.

@galak I am about to go to bed now. Is it possible for you to reproduce the issue locally? It doesn't need XCC. Just do a simple cmake on tests/boards/intel_s1000_crb/build and you should be able to readily see the issue.

galak commented:
Ok, I've run locally and see the issue, will work up a fix.

galak commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts:
```diff
@@ -35,11 +36,53 @@
 
 &spi0 {
 	status = "ok";
+	#address-cells = <1>;
```
NACK: This isn't correct for a SPI node.

The size-cells should be 0.

rgundi replied:
I get a parse error on XCC if i make it 0.

nashif replied:
> I get a parse error on XCC if i make it 0.

can you report a bug about this? Not sure why this should be XCC specific.

rgundi replied:
Well, checked again and am seeing a different behavior this time. If i keep the address-cells as 1, i get a warning when i do CMake (which i talked about here in https://github.com/zephyrproject-rtos/zephyr/pull/11850#issuecomment-444198345). If i make the address-cells as 0, this warning during CMake goes away but the subsequent "make" fails with the below error.
zephyr/drivers/flash/spi_nor.c:387: error: ‘CONFIG_FLASH_SIZE’ undeclared here (not in a function)

nashif replied:
how is this related to XCC? looks like something else is wrong here.

galak replied:
As per the dts/bindings/slave/spi,flash.yaml you should have size encoded not in reg but as a different property called 'size'.  So make this:

```
#size-cells = <0>;

flash0: flash@0 {
...
    reg = <0>;
    size = <0x400000>;
...
};

galak replied:
You will need a fixup for CONFIG_FLASH_SIZE

rgundi replied:
Thanks. Just uploaded with the modifications you suggested.

galak commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts:
```diff
@@ -35,11 +36,53 @@
 		compatible = "spi,flash";
 		label = "MX25UM512";
-		reg = <0x00>;
+		reg = <0x00 0x400000>;
```
This should just be reg = <0>;

rgundi replied:
I get a parse error on XCC if i keep this as 0. In any case, i need to specify the size of the flash. So, i need to mention 0x400000 in some manner.

galak replied:
right now dts/bindings/slave/spi,flash.yaml has a size property for this.

sathishkuttan replied:
Why can't the size be discovered via SFDP and used? I don't see a reason to limit the size to 4MB in device tree. The size should be discovered and the usage limit of 4MB should be described through the partitions definition as is being done in this PR.

galak commented on subsys/storage/flash_map/flash_map_default.c:
```diff
@@ -12,31 +12,47 @@ const struct flash_area default_flash_map[] = {
 	/* FLASH_AREA_BOOTLOADER */
 	{
 		.fa_id = 0,
+#if defined(DT_FLASH_DEV_NAME)
```
Just do this #if / #elif once like:

```
#if defined(DT_FLASH_DEV_NAME)
   #define FA_DEV_ID SOC_FLASH_0_ID
#elif (defined(CONFIG_SPI_FLASH_W25QXXDV) || defined(DT_SPI_NOR_DRV_NAME)
   #define FA_DEV_ID SPI_FLASH_0_ID
#endif
```

rgundi replied:
Sounds good. Will do.

rgundi committed:
dts: intel_s1000: Exclude sram space dedicated for mcubootloader

The starting 192kB SRAM is reserved for the mcubootloader. Exclude
this region while building an application to be loaded by mcubootloader.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

sathishkuttan commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -167,6 +145,38 @@ config GPIO_DW_0
+config SPI_NOR_SECTORS
+	default 1024
+
+choice SPI_NOR_BLOCK_ERASE_SIZE
```
According to the datasheet, the block erase size is 64K. In any case, this parameter should not be a Kconfig variable.  Instead, this parameter should be determined based on SFDP.

vanwinkeljan replied:
@sathishkuttan Note that the SPI NOR driver is used here which is intentionally using dts defines for flash dimensions over SFDP (code size & complexity). There is also a SFDP driver PR (#7561) but this one is not yet merged.

rgundi replied:
Yes. There's a different PR for that as pointed out by @vanwinkeljan. Having said that, we can use the one specified by DTS for erase size and remove the KConfig one. I wanted to do it as a different PR but now that you have pointed it out, i'll try including that change in this PR itself.

sathishkuttan commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -167,6 +145,38 @@ config GPIO_DW_0
+config SPI_NOR_SECTOR_SIZE
+	default 4096
+
+config SPI_NOR_SECTORS
```
Number of sectors should be calculated from the flash density.

rgundi replied:
You mean (Flash size)/(sector size)? Yes, that can be done.

rgundi replied:
Done

sathishkuttan commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -167,6 +145,38 @@ config GPIO_DW_0
+config SPI_NOR
+	def_bool y
+
+config SPI_NOR_JEDEC_ID
```
JEDEC ID, page size and sector size are discoverable (SFDP) and should be discovered instead of configured.

rgundi replied:
Not applicable for this PR as specified earlier.

sathishkuttan commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -167,6 +145,38 @@ config GPIO_DW_0
+	default SPI_NOR_BLOCK_ERASE_64K
+endchoice
+
+config FLASH_HAS_PAGE_LAYOUT
```
What is the concept behind "page layout"?

rgundi replied:
This is nothing but the support for different partitions (boot, slot0, slot1 and scratch area) required by mcubootloader.

sathishkuttan commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -175,6 +185,9 @@ config SPI_DW
 config SPI_0
 	def_bool y
 
+config SPI_DW_FIFO_DEPTH
```
This config should not be in the board's config file. Better not to have this this config at all -- if there is a compelling reason to have this, it should rather be in the SoC's Kconfig as this is a configuration of SPI master controller in SoC.

rgundi replied:
ok. i'll move it to soc's Kconfig.

rgundi replied:
Done

sathishkuttan commented on boards/xtensa/intel_s1000_crb/dts_fixup.h:
```diff
@@ -15,15 +15,15 @@
 	DT_SNPS_DESIGNWARE_SPI_E000_SPI_FLASH_0_LABEL
 
-#define CONFIG_SPI_NOR_SPI_SLAVE		\
+#define DT_SPI_NOR_SPI_SLAVE		\
```
Interesting choice of the name "slave". The parameter is an address, so please change to an appropriate name.

rgundi replied:
Well.. this name is generated by the python scripts for DTS, i think. Can we take this as a separate exercise? I would suggest you log a GH issue on this so that appropriate people can look into this.

sathishkuttan replied:
No. DTS generates `DT_SNPS_DESIGNWARE_SPI_E000_SPI_FLASH_0_BASE_ADDRESS` and is being redefined as `DT_SPI_NOR_SPI_SLAVE` in DTS fixup.

rgundi replied:
Done

sathishkuttan commented on drivers/flash/spi_nor.c:
```diff
@@ -338,14 +338,14 @@ static int spi_nor_configure(struct device *dev)
+	data->spi_cfg.frequency = DT_SPI_NOR_SPI_FREQ_0;
 	data->spi_cfg.operation = SPI_WORD_SET(8);
-	data->spi_cfg.slave = CONFIG_SPI_NOR_SPI_SLAVE;
+	data->spi_cfg.slave = DT_SPI_NOR_SPI_SLAVE;
```
Please change the name `slave` to something more appropriate.

rgundi replied:
see https://github.com/zephyrproject-rtos/zephyr/pull/11850#discussion_r244771987.

sathishkuttan replied:
`struct spi_config` in `include/spi.h`, defines slave as
slave is the slave number from 0 to host controller slave limit.

In this PR, is the value being assigned a slave ID or a memory address?

rgundi replied:
It is not a memory address but a slave ID (something that designates the total number of slaves from 0 to whatever number). You want it to be renamed to DT_SPI_NOR_SPI_SLAVE_ID?

sathishkuttan replied:
yes, `DT_SPI_NOR_SPI_SLAVE_ID` would be better. I got misled by the DTS generated name containing `SPI_FLASH_0_BASE_ADDRESS`

rgundi replied:
Done.

rgundi commented:
@sathishkuttan .. i have replied for your queries. Regarding SFDP, this PR doesn't address that and it's to save code space and complexity and also to address those generic flash devices which may not have SFDP supported. Will shortly upload the changes i have agreed for and will let you know about it.

rgundi committed:
flash: spi_nor: Remove Kconfig parameter for SPI_NOR_SECTORS

Number of sectors is now calculated from the flash density
i.e. (Flash size)/(sector size).

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
dts: spi_nor: Rename DT_SPI_NOR_SPI_SLAVE to DT_SPI_NOR_SPI_SLAVE_ID

The "_ID" suffix for DT_SPI_NOR_SPI_SLAVE better represents what it
is used for.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
@sathishkuttan addressed your comments

sathishkuttan commented on dts/xtensa/intel_s1000.dtsi:
```diff
@@ -31,7 +31,7 @@
 		device_type = "memory";
 		compatible = "mmio-sram";
-		reg = <0xbe000000 0x400000>;
+		reg = <0xbe030000 0x3D0000>;
```
This change affects all applications built for S1000 and the flash image generation script as well.
There should be a better way to do this - perhaps with **2 different DTS files**?

- mcuboot should be built to run from a lower (or upper) portion memory
- other applications should be built to run from remaining portion of the memory.

**Using the upper portion of memory for mcuboot might allow of easier switch between applications built pre-mcuboot-era and applications built post-mcuboot-era**

rgundi replied:
As discussed offline, have done it via the overlays.

rgundi committed:
Kconfig: spi_nor: Remove Erase configuration from Kconfig

Configuration of Erase command is specified in DTS. Hence
the corresponding changes in Kconfig are removed.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
@sathishkuttan Have addressed all your comments. Plz review.

sathishkuttan reviewed (approved):
LGTM.
@rgundi Thank you for addressing the comments!

galak commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts:
```diff
@@ -35,11 +36,54 @@
 &spi0 {
 	status = "ok";
+	#address-cells = <1>;
+	#size-cells = <0>;
```
You have #size-cells twice, I think you can probably remove both of them.

rgundi replied:
I removed one #size-cells. Couldn't figure out which was the other one.

galak commented on dts/common/mcuboot.overlay:
```diff
@@ -7,4 +7,12 @@
 		zephyr,code-partition = &slot0_partition;
 	};
+
+#ifdef CONFIG_BOARD_INTEL_S1000_CRB
```
This does not belong here, why isn't this in the board dts file?

rgundi replied:
Removed it here and moved it to the boards dts file. It will be included only when CONFIG_BOOT_MCUBOOTLOADER is defined

galak reviewed (dismissed):
1. Changes to common/mcuboot.overlay are in the wrong place
2. Minor cleanup of .dts file for #size-cells

rgundi commented:
@galak addressed your comments.

galak commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts:
```diff
@@ -14,7 +14,16 @@
+		zephyr,flash = &flash0;
 	};
+
+#ifdef CONFIG_BOOTLOADER_MCUBOOT
```
Is there a need for the ifdef?  Why not always define this?  I'm trying to remove CONFIG_ ifdef's from DTS so would prefer not to introduce a new one.

rgundi replied:
Xtensa doesn't have XIP enabled right now. As a result, the binary should be copied from flash to SRAM for execution. That holds true for MCUBoot binary too. Since this MCUBoot binary in SRAM loads the actual Zephyr binary also into the same SRAM, it cannot overwrite itself while in execution (undefined behavior mostly leading to crash). That's why we have carved out a separate non-overlapping space for MCUBoot in the SRAM. However, there are scenarios (legacy) where S1000 can run without a bootloader. Hence the use of ifdef since we don't want to disrupt them.

rgundi replied:
hmm.. is it possible to specify some overlay file in boards/xtensa/intel_s1000_crb/CMakeLists.txt and get it included in the build? Something like
if(CONFIG_BOOTLOADER_MCUBOOT)
       set(OVERLAY_FILE sram_address.dts)
endif()

Not sure how to include this in the build though (as it should be applicable only for this board).

rgundi replied:
@galak.. I see that DTC_OVERLAY_FILE can be used to specify this. If i can somehow specify the sram_address.dts file to DTC_OVERLAY_FILE, this can be handled. Do you know how to update the DTC_OVERLAY_FILE variable from boards/xtensa/intel_s1000_crb/CMakeLists.txt?

nashif replied:
you just create an overlay file ${BOARD}.overlay in the app

rgundi replied:
Yes. But the problem is it'll get added all the time. I want that file to be added only when CONFIG_BOOTLOADER_MCUBOOT is defined. Have already tried that.

rgundi replied:
Also, i would prefer to add it in boards/xtensa/intel_s1000_crb since i don't want this info to be added to all the applications. Just an overhead and not scalable.

sathishkuttan replied:
I suggest a different approach for this, {Kconfig + linker script} instead of DTS overlay.
See `BOOTLOADER_SRAM_SIZE` in `$ZEPHYR_BASE/Kconfig.zephyr` and `include/arch/arm/cortex_m/scripts/linker.ld`

Currently `Kconfig.zephyr` has a `depends on ARM` for these and will need to be extended for xtensa.

sathishkuttan replied:
LGTM

galak commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts:
```diff
@@ -35,11 +44,53 @@
 
 &spi0 {
 	status = "ok";
+	#address-cells = <1>;
```
You don't need this #address-cells (already set to 1 in dts/xtensa/intel_s1000.dtsi)

rgundi replied:
Done

galak reviewed (dismissed):
1. Remove #address-cells from spi node.
2. Why do we need adjust the memory size for MCUBOOT?  Once we transition to Zephyr aren't we free to utilize the memory that mcuboot was using?

rgundi commented:
@galak Addressed/commented on your comments.

rgundi commented:
Was hitting a dead end with the "DTC_OVERLAY_FILE" approach and hence used @sathishkuttan's suggestion. Please review.

nashif commented on boards/xtensa/intel_s1000_crb/intel_s1000_crb_defconfig:
```diff
@@ -18,6 +18,8 @@ CONFIG_3RD_LEVEL_INTERRUPTS=y
 CONFIG_CAVS_ICTL=y
 CONFIG_DW_ICTL=y
 
+CONFIG_BOOTLOADER_SRAM_SIZE=192
```
remind me, where is this 192 coming from?

rgundi replied:
This refers to the size we have reserved for the mcubootloader. 192kB.

galak commented on soc/xtensa/intel_s1000/memory.h:
```diff
@@ -6,30 +6,40 @@
 #ifndef __INC_MEMORY_H
 #define __INC_MEMORY_H
 
+#include <autoconf.h>
```
Why do you need autoconf.h here?

rgundi replied:
The autoconf.h is required for dereferencing CONFIG_BOOTLOADER_MCUBOOT.

galak reviewed (commented):
I don't know if its feasible to split this PR, so the base flash support could get merged, and than have the MCUBOOT changes as a new PR?

galak commented on soc/xtensa/intel_s1000/memory.h:
```diff
@@ -6,30 +6,40 @@
 /* L2 HP SRAM */
 #define L2_VECTOR_SIZE				0x1000
 
+#ifdef CONFIG_BOOTLOADER_MCUBOOT
```
Is this really something that should be specified by the SoC?  Seems like more a board choice.

rgundi replied:
CONFIG_BOOTLOADER_MCUBOOT is a board choice. But, the memory range is defined only in this file which is in SoC directory.

rgundi commented:
> I don't know if its feasible to split this PR, so the base flash support could get merged, and than have the MCUBOOT changes as a new PR?

These flash related changes are applicable only for MCUBoot and hence i have them as a single PR.
